connectors/es/sink.py

```diff
@@ -598,8 +598,18 @@ async def get_docs(self, generator, skip_unchanged_documents=False):
                             "doc": doc,
                         }
                     )
+                # We try raising every loop to not miss a moment when
+                # too many errors happened when downloading
+                lazy_downloads.raise_any_exception()
 
                 await asyncio.sleep(0)
+
+            # Sit and wait until an error happens
+            await lazy_downloads.join(raise_on_error=True)
+        except Exception as ex:
+            self._logger.error(f"Extractor failed with an error: {ex}")
+            lazy_downloads.cancel()
+            raise
         finally:
             # wait for all downloads to be finished
             await lazy_downloads.join()
```

Comment:
Yup, see this for example:

<img width="848" alt="image" src="https://github.com/user-attachments/assets/a40d7870-5221-408a-912e-be9baba9f192">

Reply:
In that case, the `finally` block is awaiting all of the downloads even if there's an error. But I guess `lazy_downloads.cancel()` is wiping everything, so that's okay?
(Asking because the PR title is `Cancel and fail downloads if exception happened` but it looks like it will still eventually await the downloads).

Reply:
Yeah it's intended:

Cancellation does not mean that the task is immediately dead - we need to make sure that the task has an opportunity to gracefully cancel.

For example if the cancelled task catches `asyncio.CancelledError` and does something, then it will still need to be awaited.

We use similar logic here: https://github.com/elastic/connectors/blob/8b8cb0dfffe18b6da3380a8329486157eeeab0d8/connectors/services/base.py#L177-L182

```
        for task in pending:
            task.cancel()
            try:
                await task
            except asyncio.CancelledError:
                logger.error("Service did not handle cancellation gracefully.")
```

`task.cancelI()` injects the throw of asyncio.CancelledError where it can (pretty much any `await` or `async with` statement). After that the code can catch it or not. The `except` part here is checking that task re-raised `asyncio.CancelledError` cause it could not handle it - there was no `except asyncio.CancelledError` inside task or this error was re-raised.

Hope the description makes sense, if not I can prepare an illustrative example of this :)

Reply:
That's really helpful, thanks!